tests/helpers/test_script.py

```diff
@@ -1672,74 +1071,70 @@ def wait_started_cb():
-    # TODO: REMOVE
-    if _ALL_RUN_MODES == [None]:
-        return
-
+async def test_script_mode_queue(hass):
+    """Test overlapping runs with script_mode='queue'."""
     event = "test_event"
-    events = []
-    wait_started_flag = asyncio.Event()
-
-    @callback
-    def record_event(event):
-        """Add recorded event to set."""
-        events.append(event)
-
-    hass.bus.async_listen(event, record_event)
-
-    @callback
-    def wait_started_cb():
-        wait_started_flag.set()
-
-    hass.states.async_set("switch.test", "on")
-
-    script_obj = script.Script(
-        hass,
-        cv.SCRIPT_SCHEMA(
-            [
-                {"event": event, "event_data": {"value": 1}},
-                {"wait_template": "{{ states.switch.test.state == 'off' }}"},
-                {"event": event, "event_data": {"value": 2}},
-            ]
-        ),
-        change_listener=wait_started_cb,
-        if_running="parallel",
-        run_mode="background",
-        logger=logging.getLogger("TEST"),
+    events = async_capture_events(hass, event)
+    sequence = cv.SCRIPT_SCHEMA(
+        [
+            {"event": event, "event_data": {"value": 1}},
+            {"wait_template": "{{ states.switch.test.state == 'off' }}"},
+            {"event": event, "event_data": {"value": 2}},
+            {"wait_template": "{{ states.switch.test.state == 'on' }}"},
+        ]
     )
+    logger = logging.getLogger("TEST")
+    script_obj = script.Script(hass, sequence, script_mode="queue", logger=logger)
+    wait_started_flag = async_watch_for_action(script_obj, "wait")
 
     try:
-        await script_obj.async_run()
+        hass.states.async_set("switch.test", "on")
+        hass.async_create_task(script_obj.async_run())
         await asyncio.wait_for(wait_started_flag.wait(), 1)
 
         assert script_obj.is_running
         assert len(events) == 1
         assert events[0].data["value"] == 1
 
         # Start second run of script while first run is suspended in wait_template.
-        # This should start a new, independent run.
+        # This second run should not start until the first run has finished.
+
+        hass.async_create_task(script_obj.async_run())
+
+        await asyncio.sleep(0)
+        assert script_obj.is_running
+        assert len(events) == 1
 
         wait_started_flag.clear()
-        await script_obj.async_run()
+        hass.states.async_set("switch.test", "off")
         await asyncio.wait_for(wait_started_flag.wait(), 1)
 
         assert script_obj.is_running
         assert len(events) == 2
-        assert events[1].data["value"] == 1
+        assert events[1].data["value"] == 2
+
+        wait_started_flag.clear()
+        hass.states.async_set("switch.test", "on")
+        await asyncio.wait_for(wait_started_flag.wait(), 1)
+
+        await asyncio.sleep(0)
+        assert script_obj.is_running
+        assert len(events) == 3
+        assert events[2].data["value"] == 1
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5164426)